homeassistant/components/camera/__init__.py

```diff
@@ -0,0 +1,1133 @@
+Component to interface with various cameras.
+
+The following features are supported:
+-Recording
+-Snapshot
+-Motion Detection Recording(for supported cameras)
+-Automatic Configuration(for supported cameras)
+-Creation of child entities for supported functions
+-Collating motion event images passed via FTP into time based events
+-Returning recorded camera images and streams
+-Proxying image requests via HA for external access
+-Converting a still image url into a live video stream
+-A service for calling camera functions
+
+Upcoming features
+-Camera movement(panning)
+-Zoom
+-Light/Nightvision toggling
+-Support for more devices
+-A demo entity
+-Expanded documentation
+"""
+import urllib3
+import requests
+import logging
+import time
+import math
+import datetime
+import re
+import os
+from homeassistant.helpers.entity import Entity
+from homeassistant.loader import get_component
+from homeassistant.const import (
+    ATTR_ENTITY_PICTURE,
+    HTTP_NOT_FOUND,
+    ATTR_ENTITY_ID,
+    EVENT_FTP_FILE_RECEIVED,
+    STATE_MOTION_DETECTED,
+    STATE_STREAMING,
+    STATE_ARMED,
+    STATE_ON,
+    STATE_OFF,
+    STATE_IDLE,
+    EVENT_PLATFORM_DISCOVERED,
+    ATTR_SERVICE,
+    ATTR_DISCOVERED,
+    ATTR_DOMAIN
+    )
+
+
+from homeassistant.helpers.entity_component import EntityComponent
+
+
+DOMAIN = 'camera'
+DEPENDENCIES = ['http', 'switch']
+GROUP_NAME_ALL_CAMERAS = 'all_cameras'
+SCAN_INTERVAL = 30
+ENTITY_ID_FORMAT = DOMAIN + '.{}'
+EVENT_CAMERA_MOTION_DETECTED = 'camera_motion_detected'
+EVENT_CHILD_CALLBACK_SUFFIX = '_callback'
+
+EVENT_CHANGE_RECORD = '_record_change'
+EVENT_CHANGE_MOTION = '_motion_enabled_change'
+EVENT_CHANGE_SNAPSHOT = '_snapshot_change'
+
+EVENT_CALLBACK_RECORD = '_record_callback'
+EVENT_CALLBACK_MOTION = '_motion_enabled_callback'
+EVENT_CALLBACK_SNAPSHOT = '_snapshot_callback'
+
+SWITCH_ACTION_RECORD = 'record'
+SWITCH_ACTION_MOTION = 'motion_detection'
+SWITCH_ACTION_SNAPSHOT = 'snapshot'
+
+SERVICE_CAMERA = 'camera_service'
+
+STATE_RECORDING = 'recording'
+
+DEFAULT_RECORDING_SECONDS = 30
+
+
+# The number of seconds between images before being
+# considerd a new event
+EVENT_GAP_THRESHOLD = 15
+
+# Maps discovered services to their platforms
+DISCOVERY_PLATFORMS = {}
+DISCOVER_SWITCHES = "camera.switches"
+ATTR_FRIENDLY_LOG_MESSAGE = "friendly_log_message"
+
+
+# pylint: disable=too-many-branches
+def setup(hass, config):
+    """ Track states and offer events for sensors. """
+
+    component = EntityComponent(
+        logging.getLogger(__name__), DOMAIN, hass, SCAN_INTERVAL,
+        DISCOVERY_PLATFORMS)
+
+    component.setup(config)
+
+    for entity_id in component.entities.keys():
+        entity = component.entities[entity_id]
+        entity.refesh_all_settings_from_device()
+        entity.add_child_component_listeners()
+        # entity.update_ha_state()
+
+        # This sets up and fired events for components that use the
+        # camera as a discovery platform.
+
+        data = {}
```

Comment:
This can be rewritten nicer using curly braces for everything. This example also includes extracting the name template into a constant.

```python
data = {
  ATTR_DOMAIN: DOMAIN,
  'name': RECORDER_NAME_TEMPLATE.format(entity.name),
  etc…
}
```